### PR TITLE
Fix URL sanitization and duplicate language ID

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.1.x
| Description?      | Fixes two remaining issues in routing. See below.
| Type?             | bug fix
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | See below
| Fixed issue or discussion?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/28609
| Related PRs       | 
| Sponsor company   |

### Description
- This PR fixes two issues.
- Issue 1 - With URL rewriting disabled, there was a duplicate id_lang parameter if you requested Dispatcher::createUrl with different language. Dispatcher was just appending it to the URL like `&id_lang=1` without checking if it's not already there from the request.
- Issue 2 - `FrontController::sanitizeUrl`, called from `FrontController::getAlternativeLangsUrl` was adding some extra parameters to a friendly URL so you got hreflangs with some crap. The method was using $_GET to get these parameters. We need to use the $params from the URL. :-)

### How to test
- Install [tox_langtest.zip](https://github.com/PrestaShop/PrestaShop/files/12561449/tox_langtest.zip)
- Go to `/index.php?id_lang=1&fc=module&module=tox_langtest&id_page=200&rewrite=blabla&controller=daniel&pagination=1`
- With friendly URLs enabled, see that you get a proper clean friendly URL with no extra parameters from the route. See that `pagination=1` parameter that is intentional is kept.
  - `http://localhost/81x/cs/langtest/200-blabla?pagination=1`
  - `http://localhost/81x/gb/langtest/200-blabla?pagination=1`
- With friendly URLs disabled, see that you get urls without duplicate id_lang. ✅ 
  - `http://localhost/81x/index.php?id_lang=1&pagination=1&fc=module&module=tox_langtest&id_page=200&rewrite=blabla&controller=daniel`
  - `http://localhost/81x/index.php?id_lang=2&pagination=1&fc=module&module=tox_langtest&id_page=200&rewrite=blabla&controller=daniel`

### Auto test
https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/6156334567

### CODE REVIEW

**Review by LouisAUTHIE — DISMISSED** [review]
Great one thx

**Review by boherm — DISMISSED** [review]

File: classes/controller/FrontController.php [hunk 1]
```diff
@@ -2137,18 +2137,24 @@ protected function getAlternativeLangsUrl()
...
-        $excluded_key = ['isolang', 'id_lang', 'controller', 'fc', 'id_product', 'id_category', 'id_manufacturer', 'id_supplier', 'id_cms'];
-        $excluded_key = array_merge($excluded_key, $this->redirectionExtraExcludedKeys);
-        foreach ($_GET as $key => $value) {
+        // Build a list of parameters we won't be sanitizing
+        $excludedKeys = array_merge(
+            ['isolang', 'id_lang', 'controller', 'fc', 'id_product', 'id_category', 'id_manufacturer', 'id_supplier', 'id_cms'],
+            $this->redirectionExtraExcludedKeys
+        );
+
+        // Go through each parameter we got from dispatcher and sanizize it
```
Comment (matks) [inline classes/controller/FrontController.php]:
```suggestion
        // Go through each parameter we got from dispatcher and sanitize it
```

**Review by matks — COMMENTED** [review]
One typo found

@Hlavtox You fix these problems today but maybe they will be re-introduced later by someone else. The only way to make sure we fix this once and forever is to add a Unit Test that will verify the behavior of function `createUrl()`.

Would you consider adding the unit test?

**Comment by Hlavtox** [comment]
@boherm @LouisAUTHIE @matks Can you reapprove?

Regarding the tests - I will try to have a look on it later, so this is merged in 8.1.2. :-)

**Review by matks — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]
Thanks @Hlavtox

**Comment by tups** [comment]
I just tested the PR: https://github.com/PrestaShop/PrestaShop/issues/28609#issuecomment-1717611505 and I validate :) Thanks @Hlavtox

**Comment by MatShir** [comment]
QA by dev ? We have one approve by @tups thanks to him :)
@Robin-Fischer-PS wdyt ?

**Review by M0rgan01 — APPROVED** [review]

**Comment by kpodemski** [comment]
@PrestaShop/qa-functional can we merge? already validated by a dev

**Review by aniszr — APPROVED** [review]
Hello @Hlavtox 

I tested your PR and module and it LGTM, improvements added are well verified :heavy_check_mark: 

- Friendly URL : **ON**
image


- Friendly URL : **OFF**
image

Auto Test : https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/6156334567 :heavy_check_mark: 

QA:heavy_check_mark: 

Thank you!

**Comment by nicosomb** [comment]
Thank you @Hlavtox and @aniszr!

**Pull request merged**